src/replication/ng/adapter/dockerhub/adapter.go

```diff
@@ -0,0 +1,400 @@
+package dockerhub
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/goharbor/harbor/src/common/utils/log"
+	adp "github.com/goharbor/harbor/src/replication/ng/adapter"
+	"github.com/goharbor/harbor/src/replication/ng/model"
+	"github.com/goharbor/harbor/src/replication/ng/util"
+)
+
+func init() {
+	if err := adp.RegisterFactory(model.RegistryTypeDockerHub, func(registry *model.Registry) (adp.Adapter, error) {
+		client, err := NewClient(registry)
+		if err != nil {
+			return nil, err
+		}
+
+		return &adapter{
+			client:   client,
+			registry: registry,
+			DefaultImageRegistry: adp.NewDefaultImageRegistry(&model.Registry{
+				Name:       registry.Name,
+				URL:        registryURL,
+				Credential: registry.Credential,
+				Insecure:   registry.Insecure,
+			}),
+		}, nil
+	}); err != nil {
+		log.Errorf("Register adapter factory for %s error: %v", model.RegistryTypeDockerHub, err)
+		return
+	}
+	log.Infof("Factory for adapter %s registered", model.RegistryTypeDockerHub)
+}
+
+type adapter struct {
+	*adp.DefaultImageRegistry
+	registry *model.Registry
+	client   *Client
+}
+
+// Ensure '*adapter' implements interface 'Adapter'.
+var _ adp.Adapter = (*adapter)(nil)
+
+// Info returns information of the registry
+func (a *adapter) Info() (*model.RegistryInfo, error) {
+	return &model.RegistryInfo{
+		Type:             model.RegistryTypeDockerHub,
+		SupportNamespace: true,
+		SupportedResourceTypes: []model.ResourceType{
+			model.ResourceTypeRepository,
+		},
+		SupportedResourceFilters: []*model.FilterStyle{
+			{
+				Type:  model.FilterTypeName,
+				Style: model.FilterStyleTypeText,
+			},
+			{
+				Type:  model.FilterTypeTag,
+				Style: model.FilterStyleTypeText,
+			},
+		},
+		SupportedTriggers: []model.TriggerType{
+			model.TriggerTypeManual,
+			model.TriggerTypeScheduled,
+		},
+	}, nil
+}
+
+// ValidResource checks whether a resource is valid, in DockerHub, multi-parts repo name like 'a/b/c' is not supported.
+func (a *adapter) ValidResource(resource *model.Resource) bool {
+	if resource == nil || resource.Metadata == nil || resource.Metadata.Repository == nil {
+		return false
+	}
+
+	if len(strings.Split(resource.Metadata.Repository.Name, "/")) != 1 {
+		return false
+	}
+
+	return true
+}
+
+// ConvertResourceMetadata converts the namespace and repository part of the resource metadata
+// to the one that the adapter can handle
+func (a *adapter) ConvertResourceMetadata(meta *model.ResourceMetadata, namespace *model.Namespace) (*model.ResourceMetadata, error) {
+	return meta, nil
```

Comment:
how to process full repository path like `a/b/c:tag`?

Reply:
For the moment, only Habor -> DockerHub, this suffices, for other cases, let's consider it later after  https://github.com/goharbor/harbor/issues/7361